135579 navigation bundle bem verwenden

megatronCGN opened the pull request:
Das alte ni-na-nap Klassen-Pattern muss sterben weil schlecht zu lesen, viel Overhead, nicht dokumentiert, überholt. Üblicherweise überschreiben wir (Team Grün) die Navigation(Blöcke) ein mal komplett auf Projekt-Ebene mit BEM Klassen. Die Arbeit möchte ich ins in Zukunft ersparen und eine Version 5 des Bundles per default damit ausstatten.

megatronCGN committed:
Update navigationBlocks.html.twig (Case 135579)

- Entferne ni-na-nap Klassen-Overhead
- Etabliere Klassenstruktur nach BEM
- Ermögliche Prefixen der Klassen
- Definiere Wildcard Klassen Blocks auf allen Elementen für Kontext-abhängigen Bedarf an weiteren Klassen (z.B. JS-Hooks)

megatronCGN committed:
Update navigationBlocks.html.twig (Case 135579)

- Vereinfache Benennung der Prefix Blocks

megatronCGN committed:
Update Breadcrumbs (Case 135579)

- Nutze Use Pattern um Blöcke überschreibbar zu machen
- Ermögliche Prefixes
- Ermögliche Kontext-abhängige zusätzliche Klassen

megatronCGN committed:
Update CHANGELOG.md (Case 135579)

- Dokumentiere BC-Breaks und Änderungen

MalteWunsch commented on src/Resources/views/Navigation/navigationBlocks.html.twig:
```diff
@@ -2,29 +2,43 @@
         {% set visibleNodes = root.children | filter(node => node.visible) %}
         {% if visibleNodes is not empty %}
-            {{ block('navigation_list') }}
+            <nav class="{{ navigation_base_class }} {% block navigation_additional_class %}{% endblock %}">
```
Ist glaube ich ein weiterer BC-Break, weil mit Version 4 erwartet wurde, das `<nav>` im Projekt zu schreiben (weiß nicht warum, gibts da mehr Flexpoints?). Wäre dann nach dem Upgrade ein doppeltes `<nav>` und man müsste nicht nur CSS anpassen.

polarbirke replied:
Sehe das wie Malte. Wenn wir die `<nav>` Elemente mit reinnehmen wollen, brauchen wir noch einen weiteren flexpoint für `aria-label` um die Navigation Landmarks in AT unterscheidbar zu machen ("main", "breadcrumbs", "legal" sind typische englische Bezeichnungen, für die ich keine prägnanten deutschen Übersetzungen parat habe).

megatronCGN replied:
Guter Punkt

MalteWunsch reviewed (commented):
Find ich ne gute Idee.

> Das alte ni-na-nap Klassen-Pattern muss sterben weil [...] nicht dokumentiert

Ist es sinnvoll, die jetzigen CSS-Klassen/den Aufbau einer Navigation anhand eines Beispiels in der README zu dokumentieren oder ist das für Team Grün bloody selbsterklärend (für mich nicht)?

polarbirke commented:
https://github.com/webfactory/WebfactoryNavigationBundle/blob/8098186d13bd4ace7ce12f111d57b427df8db63d/src/Resources/views/Navigation/navigationBlocks.html.twig#L77-L79

Vorschlag: wir machen auch hier einen BC Break und wechseln für aktive Nodes von `<span>` zu `<a aria-current="page">`. So wird es überall empfohlen, weil es explizitere Information transportiert ("ich bin die aktuelle Seite" stattt "ich bin kein Link"). 

Wenn ich mich bei meiner kurzen Recherche gerade nicht vertue, ist es erlaubt & korrekt, mehr als 1 `aria-current="page"` auf einer Seite zu haben (z.B. Hauptnavigation + Breadcrumb).

Wir könnten dann auch (endlich) `{{ block('navigation_text_class') }}` zu `{{ block('navigation_link_class') }}` umbenennen (und die Klasse natürlich auch). Das finde ich a) intuitiver und b) erlaubt es uns, im Zweifelsfall noch ein `<span class="nav__text">` in die Links zu verschachteln. Auch das wäre ein weiterer _breaking_ Vorschlag.

megatronCGN commented:
> Find ich ne gute Idee.
> 
> > Das alte ni-na-nap Klassen-Pattern muss sterben weil [...] nicht dokumentiert
> 
> Ist es sinnvoll, die jetzigen CSS-Klassen/den Aufbau einer Navigation anhand eines Beispiels in der README zu dokumentieren oder ist das für Team Grün bloody selbsterklärend (für mich nicht)?

Allen in Team Grün sollte das klar sein. Die BEM Doku ist auch ein besserer Startpunkt. Ich weiss nicht, ob es Sinn macht, den Aufbau zu dokumentieren, denn den sieht man ja im Inspektor. Eine Anleitung zum Stylen wäre zu umfangreich denke ich. Würde dir das fehlen?

megatronCGN commented:
> https://github.com/webfactory/WebfactoryNavigationBundle/blob/8098186d13bd4ace7ce12f111d57b427df8db63d/src/Resources/views/Navigation/navigationBlocks.html.twig#L77-L79
> 
> Vorschlag: wir machen auch hier einen BC Break und wechseln für aktive Nodes von `<span>` zu `<a aria-current="page">`. So wird es überall empfohlen, weil es explizitere Information transportiert ("ich bin die aktuelle Seite" stattt "ich bin kein Link").
> 
> Wenn ich mich bei meiner kurzen Recherche gerade nicht vertue, ist es erlaubt & korrekt, mehr als 1 `aria-current="page"` auf einer Seite zu haben (z.B. Hauptnavigation + Breadcrumb).
> 
> Wir könnten dann auch (endlich) `{{ block('navigation_text_class') }}` zu `{{ block('navigation_link_class') }}` umbenennen (und die Klasse natürlich auch). Das finde ich a) intuitiver und b) erlaubt es uns, im Zweifelsfall noch ein `<span class="nav__text">` in die Links zu verschachteln. Auch das wäre ein weiterer _breaking_ Vorschlag.

Würdest du das zum Default machen? Ich glaube ich habe das zumindest schon mehrfach händisch so gebaut, weil ich ein zusätzliches Element brauchte.

MalteWunsch commented:
> Allen in Team Grün sollte das klar sein. Die BEM Doku ist auch ein besserer Startpunkt. Ich weiss nicht, ob es Sinn macht, den Aufbau zu dokumentieren, denn den sieht man ja im Inspektor. Eine Anleitung zum Stylen wäre zu umfangreich denke ich. Würde dir das fehlen?

Mir persönlich überhaupt nicht! Ich werf euch am Liebsten ein `{{ navigation_tree(root = {"webfactory_pages.navigation_tags": "startseite"}) }}` über den Zaun und warte dann ein paar Tage, bis das rumms-zischel-krach zu Ende ist und alles gut aussieht.

Wie viel Doku ihr in Team Grün haben wollt, da will ich gar nicht reinreden. Vielleicht wäre @DerMeier s Meinung interessant, der sich am Ehesten in einen Anfänger reinversetzen kann?

megatronCGN committed:
Update navigationBlocks.html.twig, breadcrumbsBlocks.html.twig, CHANGELOG.md (Case 135579)

- Update bzw. füge neue Blocks für Links und Text hinzu
- Ersetze span-Element des aktive Navigationspunkt durch a-Element mit Aria-Attribut
- Dokumentiere BC-Breaks und Änderungen

polarbirke commented:
> Würdest du das zum Default machen? Ich glaube ich habe das zumindest schon mehrfach händisch so gebaut, weil ich ein zusätzliches Element brauchte.

Ich vermute, du meinst das in den Link verschachtelte `<span>`? Ich habe grad kein konkretes Beispiel im Kopf, wo ich es brauchen würde und deswegen das Gefühl, dass es eher eine Ausnahme ist und kein default sein sollte.

Vielleicht habe ich schon zu lange keine Navigation mehr gebaut? Wo hast du es gebraucht?

megatronCGN committed:
Update navigationBlocks.html.twig (Case 135579)

- schreibe Levels auf die navigation-text Klassen

polarbirke commented on src/Resources/views/Navigation/breadcrumbsBlocks.html.twig:
```diff
@@ -0,0 +1,34 @@
+                    {% if node.caption|trim is not empty and ((node.breadcrumbsVisible is defined and node.breadcrumbsVisible) or node.visible) %}
+                        <li class="{{ block('prefix_final') }}breadcrumbs__item {% block breadcrumbs_item_class %}{% endblock %}">
+                            <a class="{{ block('prefix_final') }}breadcrumbs__link {% block breadcrumbs_link_class %}{% endblock %}"
+                                {%- if not node.activeNode -%}
+                                    href="{{ node.url }}"
+                                {%- else -%}
+                                    aria-current="page"
+                                {%- endif  -%}>
```
Halt, Moment. Das sollte sein:

```
href="{{ node.url }}" {% if node.activeNode -%}aria-current="page"{%- endif  %}>
```

weil Links ohne `href` nicht fokussierbar sind.

megatronCGN replied:
Ich dachte immer das wäre der Sinn des `<span>`

megatronCGN replied:
... aber natürlich bekommen user für die die Info interessant ist sie nicht mit, wenn das element nicht fokussierbar ist.

polarbirke replied:
Ja, das war schon wf best practice als ich 2007 angefangen habe. Aber aus heutiger Sicht dank ARIA nicht mehr sinnvoll.

polarbirke commented on src/Resources/views/Navigation/navigationBlocks.html.twig:
```diff
@@ -45,38 +59,35 @@
     {%- if not node.isActiveNode and block('navigation_url') -%}
-        <a href="{{ block('navigation_url') }}" class="{{ block('navigation_text_class') }}">{{ block('navigation_caption') }}</a>
+        <a href="{{ block('navigation_url') }}" class="{{ block('navigation_link_class') }}">{{ block('navigation_caption') }}</a>
     {%- else -%}
-        <span class="{{ block('navigation_text_class') }}">{{ block('navigation_caption') }}</span>
+        <a aria-current="page" class="{{ block('navigation_link_class') }}">{{ block('navigation_caption') }}</a>
```
Hier auch nicht `href` oder `aria-current` sondern immer `href` und `if node.isActiveNode` zusätzlich `aria-current`.

megatronCGN committed:
Update navigationBlocks.html.twig, breadcrumbsBlocks.html.twig (Case 135579)

- href-attr ist mandatory für das Link-Element, damit es fokussierbar ist

DerMeier commented:
> Wie viel Doku ihr in Team Grün haben wollt, da will ich gar nicht reinreden. Vielleicht wäre @DerMeier s Meinung interessant, der sich am Ehesten in einen Anfänger reinversetzen kann?

Also ein Doku für BEM brauche ich nicht. Würde das dann so machen wie sh das vorgeschlagen hat (mit dem Inspektor).

megatronCGN committed:
Update changelog.md (Case 135579)

- Fixe Typos
- Verbessere Formulierungen

mpdude commented:
Mir scheint ich komme ein bisschen spät zur netten kleinen Party hier...

Inhaltlich kann ich es nicht beurteilen, da ist Team 🟩 viel näher dran und alle, die zuletzt an Projekten mit diesem Bundle gearbeitet haben. Ich verstehe sehr gut, dass dass diese CSS-Klassen (so sehr sich jemand mal was dabei gedacht hat) und auch das `power_set` ([Potenzmenge](https://de.wikipedia.org/wiki/Potenzmenge)) gruselig sind und die Welt ohne sie ein besserer Ort wäre.

Als Teil der älteren Generation muss ich aber zu bedenken geben, dass meine mentale Kapazität langsam begrenzt ist und ich BC Breaks und Major Versions immer mit Schrecken zur Kenntnis nehme. Daher ein paar Denkanstöße:

* Wenn wir das machen, dann zwingen wir uns über kurz oder lang in _jedem_ Projekt, das dieses Bundle einsetzt, auch auf die neue Navigation, Struktur, BEM – whatever wir hier tun – zu wechseln. (Wenn irgendwann eine aktuelle Symfony-Version nur noch zu haben ist, wenn man `webfactory/navigation-bundle` `^5.0` oder größer einsetzt, wird der Druck groß.)
* So ein Upgrade wird unschön, weil es wahrscheinlich weder jemand von Team 🟦 noch 🟩 alleine machen kann, sondern das Bundle-Upgrade mit einem großen CSS/Twig-Umbau verwoben ist. Es wird ein blödes Wechselspiel zwischen den 🟩/🟦 geben.
* Wir reden über einen blöden HTML/CSS-Umbau, der wahrscheinlich durch keinerlei existierende Tests unterstützt wird und/oder vermutlich nur mit Tests abzusichern wäre, die HTML-Struktur und/oder visuelle Regression testen.
* Wenn die Version 4.x noch das "alte" und die Version 5.0 nur das "neue" HTML kann, dann kann man nicht mal einfache "Wackeltests" machen. Damit meine ich Tests, in denen man einfach abwechselnd beide Versionen ausprobieren kann, indem man nur eine Zeile ein- oder auskommentiert oder sogar über zwei URLs beide Versionen parallel aufrufen kann. Man kann immer nur mit einem Paket Up-/Downgrade wechseln und _nur_ das alte _oder nur_ das neue haben.

Daher mein Vorschlag:

Findet ihr einen Weg, das "neue" HTML/CSS/whatever _zusätzlich_ zum alten in dieses Bundle zu packen? 

Das muss ja nicht heißen, dass das alte noch aktiv weiterentwickelt oder gepflegt wird, aber vielleicht kann es feature-frozen "da bleiben" (oder in einem `legacy-navigation-theme` zusätzlich nachinstallierbar sein)?

Wir könnten den BC break vermeiden, wenn das neue Theme (wenn ich das mal so nennen darf) ergänzend im Bundle dazu kommt und man z. B. über eine Konfigurationseinstellung, ein `{% extends ... %}` in Twig oder wie auch immer auswählen kann, welche Version man nutzen möchte. Wenn der BC Break irgendwann nur darin besteht, welche Version by default aktiv ist, ist alles halb so schlimm, wenn man leicht zwischen beiden Varianten wechseln kann.

Also: Möchtet Ihr Euch wirklich in die Situation zwingen, in allen Projekten diesen Umbau machen zu müssen?

polarbirke commented:
The breaking changes are coming FROM INSIDE THE HOUSE 😱

sebastiankugler commented:
Finde @mpdude s Vorschlag sehr gut. Wenn man in einer neuen Version des Bundles erstmal alt und neu nebeneinander anbieten würde (oder über ein `legacy-navigation-theme` zusätzlich installierbar), könnten wir alle Projekte problemlos weiter updaten und dann sukzessive auf das neue Theme umstellen. Wenn die Umstellung in allen oder den meisten Projekten erfolgt ist, könnte der Support fürs Legacy-Theme dann in einer neuen Major entfernt werden (aber idealerweise nicht vorher).

Projekte, bei denen wir das Theme ohnehin schon überschrieben haben, wären vermutlich aber gar nicht von dem BC break betroffen, würden aber auch gar nicht von der neuen Version profitieren? 

Wichtig ist vermutlich, dass wir im Bundle überhaupt erstmal ein modernes Theme anbieten, sodass wir nicht gezwungen sind, in jedem Projekt eins anzulegen.

megatronCGN commented:
@polarbirke und ich hatten uns aktiv dagegen entschieden das neue Theme neben das alte zu legen, damit man eben irgendwann aufhört das alte mitzuschleppen auch wenn es schmerzhaft ist. Wie realistisch ist es, dass man im Bulk alle Projekte mit alten Theme auf das neue umstellen muss? Zumal es "nur" kosmetische Änderung im weitesten Sinne sind. Einen einfachen Wackeltest wird es auch mit 2 parallelen Themes nicht geben, dafür sind die Änderungen einfach zu umfangreich. 
Allerdings war die Entscheidung pro BC-Breaks durchaus ein knappe und ich sehe auch den Vorteil davon, sich das nicht aufbürden zu müssen. Andererseits sehe die Chancen dann auch höher, dass wir niemals upgraden.

megatronCGN commented:
@sebastiankugler Projekte die das aktuelle Theme überschreiben sind die ersten, die ich anpassen würde (bei nächster Gelegenheit). Die Folgen mehr oder weniger schon diesem Pattern.

megatronCGN committed:
Erzeuge Kopien der alten Templates als Vorbereitung zum Merge mit BEM-Templates (Case 135579)

megatronCGN committed:
Merge branch '135579_Navigation-Bundle-BEM-verwenden--temp' into 135579_Navigation-Bundle-BEM-verwenden

megatronCGN committed:
Suffixe neue Template-Dateien, damit sie mit alten Templates koexistieren können (Case 135579)

megatronCGN committed:
Stelle alte Templates wieder her (Case 135579)

megatronCGN committed:
Stelle NavigationExtension.php mit power_set Methode wieder her (Case 135579)

megatronCGN committed:
Update CHANGELOG.md (Case 135579)

megatronCGN committed:
Korrigere Pfade (Case 135579)
